gridded_etl_tools/utils/extractor.py

```diff
@@ -10,75 +10,64 @@
-
-    Parameters
-    ----------
-    batch_processor
-        Function which will be called repeatedly, being passed a new entry in the batch each time it's called
-    batch
-        A 2D list of lists (or any sequential type) containing arguments to be forwarded to the given batch
-        processor
-
-    Returns
-    -------
-    bool
-        `True` if any of the batch calls returns `True`, `False` otherwise
-    """
-    # Success remains Trueuntil the first unsuccessful request is completed
-    success = True
+class Extractor(ABC):
+
+    def __init__(self, dm: dataset_manager.DatasetManager, concurrency_limit: int = 8):
+        self.dm = dm
+        self.semaphore = asyncio.Semaphore(concurrency_limit)
+
+    def pool(self, batch: typing.Sequence[typing.Sequence]) -> bool:
+        """
+        Executes a batch of jobs concurrently using asyncio.
 
-    # Time the download
-    start_downloading = time.time()
+        Args:
+            batch (typing.Sequence[typing.Sequence]): A sequence of job arguments.
 
-    # Thread count is one less than the amount of CPU available, unless there is only 1 CPU
-    thread_count = max(1, multiprocessing.cpu_count() - 1)
+        Returns:
+            bool: True if all of the jobs succeeded, False otherwise.
+        """
+        nest_asyncio.apply()
+        coros = [asyncio.to_thread(self.request, *job_arg) for job_arg in batch]
+        results = asyncio.run(self.async_gather_with_semaphore(coros))
+        return all(results)
+
+    async def async_gather_with_semaphore(self, coros: list[typing.Coroutine]) -> list[bool]:
+        """
+        Asynchronously executes a list of coroutines with a semaphore.
 
-    # run downloads
-    if len(batch) > 0:
-        log.info(f"Submitting request for {len(batch)} datasets in parallel using {thread_count} threads.")
+        Args:
+            coros (list[typing.Coroutine]): A list of coroutines to execute.
 
-        # download in parallel
-        with multiprocessing.pool.ThreadPool(processes=thread_count) as pool:
-            for result in pool.starmap(batch_processor, batch):
-                if not result:
-                    success = False
+        Returns:
+            list[bool]: A list of the results of the coroutines
+        """
 
-        log.info(f"Downloading took {(time.time() - start_downloading) / 60:.2f} minutes")
-    else:
-        log.info("No requests prepared to be submitted, please check batch_processor is functioning properly.")
-        success = False
+        async def sem_coro(coro):
+            async with self.semaphore:
+                return await coro
 
-    # If every request returned False, log a message
-    if not success:
-        log.info("At least one request for data was unsuccessful.")
+        return await asyncio.gather(*(sem_coro(c) for c in coros))
 
-    return success
+    @abstractmethod
+    def request(self, *args, **kwargs) -> bool:
```

Comment:
Should include a short docstring detailing expectations for child implementations.